### ISSUE TITLE
Fix memory leak in CommonMark implementation

### ISSUE THREAD

**Issue by Witiko** [Witiko/markdown#308]
In https://github.com/Witiko/markdown/pull/226#issuecomment-1599641634, we have identified a memory leak in our implementation of CommonMark. https://github.com/Witiko/markdown/pull/226#issuecomment-1602414441 has shown that large data structures are created at every call of `markdown.new()` and not freed by garbage collection. The issue has been hotfixed in https://github.com/Witiko/markdown/pull/226/commits/1c4b844a3d51f06446a4b138dc9ae3e8d6e0353d by adding option `singletonCache` that caches the result of calling `markdown.new(options)` to an LRU cache of size 1 keyed by `options`, but documents with a large number of Markdown fragments typeset with different options are still affected.

### Acceptance criteria

First, navigate to the root folder of the git repository and save the following script to a file named `memory-leak.lua`:

``` lua
#!/usr/bin/env texlua
local ran_ok, kpse = pcall(require, "kpse")
if ran_ok then kpse.set_program_name("luatex") end
local markdown = require("markdown")

for _ = 1, 100 do
  markdown.new({singletonCache = false})("Hello *world*!")
end
```

Next, create executable files named `bisect.sh`, `build-docker-image.sh`, `run-test.sh`, and `remove-docker-image.sh`, as described in https://github.com/Witiko/markdown/pull/226#issuecomment-1602414441, section Experimental setup.

Finally, run the command `./bisect.sh`. Here is the expected result:

```
$ ./bisect.sh
No problems were detected, success.
```

Afterwards, you may want to clear the build cache of Docker:

```
$ docker builder prune
```

**Comment by Witiko** [comment]
Here is the average memory leaked for different calls, drilled down to the affected call sites:

- `M.new()`: Average memory leak based on 23 runs: 41.65MiB
    - `M.writer.new()`: Average memory leak based on 23 runs: 0.30MiB
    - `M.reader.new()`: Average memory leak based on 25 runs: 8.84MiB
    - `reader->finalize_grammar()`: Average memory leak based on 21 runs: 30.76MiB

### Experimental details

First, I added the following sentinel code into the Lua parts of `markdown.dtx`:

``` lua
-- Log RAM at the beginning of the test.
os.execute("free -m | head -2 | awk '{ print $1,$2,$3,$4 }' | sed 's/ shared//'")
```

``` lua
-- Log RAM at the end of the test.
os.execute("free -m | head -2 | tail -1 | awk '{ print $1,$2,$3,$4 }'")
```

Then, I created a file `analyze-profile.py` with the following contents:

``` py
import statistics
import sys

deltas = []
previous_used = None
for line in sys.stdin:
    if line.startswith('Mem: '):
        _, total, used, free = line.split()
        used = int(used)
        if previous_used is None:
            previous_used = used
            continue
        delta = used - previous_used
        deltas.append(delta)
        previous_used = None
print(f'Average memory leak based on {len(deltas)} runs: {statistics.mean(deltas):.2f}MiB')
```

Finally, I measured the average memory leaked by running `./bisect.sh 2>/dev/null | python3 analyze-profile.py`.

**Comment by Witiko** [comment]
@lostenderman Above, I have narrowed down the memory leaks to tens of lines of code. However, I have mainly focused on discovery so far and I don't have an exact idea about the nature of the memory leak yet. If you could spare the time, I would appreciate it if you could glance over the affected code and brainstorm about the possible causes.

**Comment by Witiko** [comment]
I had a short brainstorming session about the issue with AI-enabled Bing, which suggested the following: ``One possible cause is that you are defining recursive patterns that create reference cycles between them.'' I can see how this could be a problem and how this could prevent garbage collection.

**Comment by Witiko** [comment]
> 10.48MiB > 4.92MiB is likely due to delayed garbage collection

To improve the measurement accuracy, I prefixed both parts of the sentinel code from https://github.com/Witiko/markdown/issues/308#issuecomment-1675939694 with `collectgarbage("collect")`. As an unintended result, this has caused the memory leak to disappear. Apparently, there is no memory leak, just the garbage collection is too slow to reclaim the memory allocated for temporary objects before an out-of-memory event. I used this theory to construct the fix in https://github.com/Witiko/markdown/pull/318.
